### PR TITLE
Backport [1/5]: IO system refactorings

### PR DESCRIPTION
**Pull request by rapgenic**
This is the first set of patches, mainly regarding input and output functions.

It can be resumed in the following points:

* the io override/restore mechanism has been removed: the output must always be sent to jupyter even if a cell isn't executing (this is important for widgets).
* a new function for registering symbols/functions in octave context has been added
* input and output classes were split in two files and put in their own namespace for tidiness

### CODE REVIEW

**Comment by AntoinePrv** [comment]
Hi @rapgenic, sorry I missed that. Don't hesitate to ping me or request me as reviewer. I'll get on it.

**Comment by rapgenic** [comment]
Don't worry, probably my fault for not adding you as a reviewer.

Anyway it allowed me to test it for a longer time, which is a good thing

File: include/xeus-octave/output.hpp [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with xeus-octave.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef XEUS_OCTAVE_OUTPUT_HPP
+#define XEUS_OCTAVE_OUTPUT_HPP
+
+#include <mutex>
```
Comment (AntoinePrv) [inline include/xeus-octave/output.hpp:23]:
```suggestion
#include <mutex>
#include <string>
```

File: include/xeus-octave/utils.hpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+#include <octave/interpreter.h>
+#include <octave/ov-builtin.h>
+#include <octave/ov-fcn-handle.h>
+#include <octave/ov-null-mat.h>
+#include <octave/ov.h>
+
+#include <iostream>
+
+#include "xeus/xguid.hpp"
+#include "xtl/xoptional.hpp"
```
Comment (AntoinePrv) [inline include/xeus-octave/utils.hpp]: Suggestion: I prefer keeping `"` for this project headers
```suggestion
#include <iostream>

#include <octave/interpreter.h>
#include <octave/ov-builtin.h>
#include <octave/ov-fcn-handle.h>
#include <octave/ov-null-mat.h>
#include <octave/ov.h>
#include <xeus/xguid.hpp>
#include <xtl/xoptional.hpp>
```

File: src/xoctave/display.cpp [hunk 3]
```diff
@@ -17,44 +17,47 @@
...
+ */
 octave_value_list display_data(octave_value_list const& args, int /*nargout*/)
 {
+  // Agruments check
   if (args.length() < 1 || args.length() > 2)
     print_usage();
 
-  nlohmann::json data;
+  // Get the data object
+  nl::json data(nl::json::value_t::object);
```
Comment (AntoinePrv) [inline src/xoctave/display.cpp:47]: Is this equivalent to `nl::json data = nl::json::object()`?

**Review by AntoinePrv — APPROVED** [review]
I'm probably not as familiar with the code as you, but looks good. Thanks!

File: include/xeus-octave/utils.hpp (hunk 2 above)
Comment (rapgenic) [reply include/xeus-octave/utils.hpp]: Yes that's from the past when xeus was in tree

File: src/xoctave/display.cpp (hunk 3 above)
Comment (rapgenic) [reply src/xoctave/display.cpp:47]: I'm not sure, it probably is, without a copy. I'm sure however that the line is creating an empty json object (as per documentation), which is the intended behaviour.

**Comment by rapgenic** [comment]
Sorry for letting this slide for so long, I had a very busy month...

**Pull request merged**